### ISSUE TITLE
Test times out unexpectedly, despite expected behavior otherwise

### ISSUE THREAD

**Issue by dwschrashun** [just-jeb/jest-marbles#162]
Hi there, @clavery63 and I have noticed some behavior that we believe is a bug, resulting in tests timing out when they should complete successfully.

I've reproduced here https://codesandbox.io/s/suspicious-ardinghelli-0yt2n

```
import { cold } from "jest-marbles";

describe("timeout error", () => {
  test("times out without completing observable", done => {
    const fakeActions = cold("t|", { t: true });

    fakeActions.subscribe({
      next: value => {
        console.log("value received:", value);
        expect(value).toEqual(true);
      },
      complete: () => {
        console.log("observable complete");
        done();
      }
    });
  });
});
```

Based on the way the logs appear, and our understanding of the tools involved, we'd expect the test to pass, but instead it times out.

If this is expected behavior, please let us know if you have thoughts on how to go about testing this sort of thing effectively. Thanks!

**Comment by dwschrashun** [comment]
Perhaps worth noting that substituting a standard Observable for the marble based Observable works as expected:

```
import { of } from "rxjs";

describe("timeout error", () => {
  test("times out despite logging correctly", done => {
    const fakeActions = of(true);

    fakeActions.subscribe({
      next: value => {
        console.log("value received:", value);
        expect(value).toEqual(true);
      },
      complete: () => {
        console.log("observable complete");
        done();
      }
    });
  });
});
```

**Comment by just-jeb** [comment]
Can you elaborate on your specific use case? I understand that you're providing here a minimal reproduction but it is hardly a real use case (testing the cold observable you defined) so I'm not sure whether it should be expected behavior or not.
Normally the test flow would be as following:
1. You define an input observable (or multiple input observables) with `cold` or `hot`. 
2. You apply the operator (for example `map`) that you want to test and it returns another observable that we'll call `outcome`. 
3. You define the expected outcome (with `cold` or `hot` or just a marble string) 
3. You use `expect(outcome).toBeObservable(expectedOutcome)` or `expect(outcome).toBeMarble(expectedOutcome)` to match the results.

**Comment by dwschrashun** [comment]
Here's a more robust example. My hope and expectation is that we are able to treat Observables created with jest-marbles cold and hot wrappers largely as we would rxjs Observables, but these results seem to suggest that is not the case. Let me know what you think. 

```
import { cold } from "jest-marbles";
import { map } from "rxjs/operators";
import { of } from "rxjs";

const mapStreamWith = mapper => stream => stream.pipe(map(mapper));

describe("example with spy", () => {
  test("test times out", done => {
    const fakeStream = cold("t|", { t: "fake-value" });
    const mapSpy = jest.fn();
    mapSpy.mockReturnValue("mapped-value");
    const streamMapper = mapStreamWith(mapSpy);

    streamMapper(fakeStream).subscribe({
      complete: () => {
        expect(mapSpy.mock.calls.length).toEqual(1);
        done();
      }
    });
  });
  test("test does not time out", done => {
    const fakeStream = of({ t: "fake-value" });
    const mapSpy = jest.fn();
    mapSpy.mockReturnValue("mapped-value");
    const streamMapper = mapStreamWith(mapSpy);

    streamMapper(fakeStream).subscribe({
      complete: () => {
        expect(mapSpy.mock.calls.length).toEqual(1);
        done();
      }
    });
  });
});

```

**Comment by just-jeb** [comment]
If I understand it right you're trying to make sure your `mapStreamWith` method works correctly.  
I'd test it without subscriptions and async. The following code should cover your test case:
```js
import { cold } from "jest-marbles";
import { map } from "rxjs/operators";
import { of } from "rxjs";

const mapStreamWith = mapper => stream => stream.pipe(map(mapper));

describe("example with spy", () => {
  test("test that mapper applied correctly", done => {
    const fakeStream = cold("t|", { t: "fake-value" });
    const mapSpy = jest.fn();
    mapSpy.mockReturnValue("mapped-value");
    const streamMapper = mapStreamWith(mapSpy);
    const result = streamMapper(fakeStream);
    const expectedResult = cold("r|": {r: "mapped-value"});
    expect(result).toBeObservable(expectedResult);
  });
})
```
Or with `toBeMarble`:
```js
describe("example with spy", () => {
  test("test that mapper applied correctly", done => {
    const fakeStream = cold("t|");
    const mapSpy = jest.fn();
    mapSpy.mockReturnValue("v");
    const streamMapper = mapStreamWith(mapSpy);
    const result = streamMapper(fakeStream);
    expect(result).toBeMarble("v|");
  });
})
```

In general, the whole marble-testing paradigm is intended to make your async observable operators test synchronous and simplify them so won't have to subscribe.  

Now, if you still want to make it work in your way can please you try one thing?
Try to run your tests, but don't use `done` (don't call it from within `subscribe` callback).  
The reason is that `TestScheduler` is currently flushed after each test and if you call `done` there is no AFTER for this test as it times out. If your test stays synchronous (without `done`) then the subscription flow will happen after the test and it would still fail if something is wrong.

I'll explore the possibility of flushing the scheduler after each `expect` (which should solve this specific problem) but it will probably introduce new issues.

**Comment by dwschrashun** [comment]
Thanks for the response. It seems like the issue might be that jest-marbles only really supports explicitly functional synchronous tests that take an input stream and then assert against an output stream. If we need to test something that is more of a side effect in this process, for example, that a mock Subject passed into a Redux Observable epic will receive values based on the epic's action Observable, perhaps we need to use the built in RxJS testing mechanisms or a different library to mock the streams?

If so, it might be nice to add a sentence or two in the readme about this sort of test pattern not being supported.

Now, to step through the merge operator and try to figure out why I'm still having problems with my tests, depending on the order in which I merge my Observables, even when I pull out jest-marbles. Thanks!

**Comment by just-jeb** [comment]
Yeah, it seems that testing side effects is currently a bit problematic. I have a few things in mind (such as adding `toHaveSideEffect` matcher), will see what can be done.  
There is a workaround for you until it's done though. You can pipe to the tested stream and `tap` your assert:  
```
    const mapStreamWith = (mapper: any) => (stream: any) => stream.pipe(map(mapper));
        const fakeStream = cold("t|");
    const mapSpy = jest.fn().mockReturnValue('r');
    const streamMapper = mapStreamWith(mapSpy);
    const testee = streamMapper(fakeStream).pipe(tap(() => {
      expect(mapSpy.mock.calls.length).toEqual(1);
    }))
    expect(testee).toBeMarble('r|');
```

**Comment by just-jeb** [comment]
I am gonna close this in favor of #112 (essentially a duplicate). Please follow on that issue to get updates.

**Comment by just-jeb** [comment]
Please use [`toSatisfyOnFlush`](https://github.com/just-jeb/jest-marbles#tosatisfyonflush), available since 2.5.0.  
Let me know if that helps.

**Comment by hanvyj** [comment]
`toSatisfyOnFlush` worked great for me. Thanks!
